rfcs/20201121-keras-model-fit-ps.md

```diff
@@ -368,11 +368,54 @@ Since the workers will sync every epoch anyway, fetching the remote values incur
+
+**Cons:**
+
+- There is no reasonably performant way to support batch-level `Callback`s. This is because to support batch-level `Callback`s, we would need to sync after every step. When a step is defined as a single batch on a single worker, this means that only one worker will ever be non-idle when batch-level `Callback`s are passed. We cannot even rely on the `steps_per_execution` parameter, because this parameter only controls how many batches are run within one `tf.function`, but the `tf.function` is still only run on one worker. Meaning even with a high `steps_per_execution`, batch-level `Callback`s would cause all but one worker to be idle at any given time. Because of this, if we go with this mental model, we should error out when a user attempts to pass batch-level `Callback`s to `Model.fit`
+- If a user passes a `Dataset` with `D` batches to each worker, they might expect that setting `steps_per_epoch=D` will visit every batch of the `Dataset`. However, only the first `D/W` batches will be visited with this mental model.
+
+**Option 2: A `step` is one batch on every worker**
+
+With this mental model, every time `Model.train_function` is called, it schedules one batch to execute on each worker. This means that if there are `W` workers, passing `steps_per_epoch=100` will actually run `100 * W` batches of training, with each worker seeing `100` batches.
+
+**Pros:**
+
+- Batch-level `Callback`s can be supported with this approach. We can sync every step, and we can rely on the `steps_per_execution` parameter to ensure that the step time is large enough so that this syncing is not prohibitively expensive.
+
+**Cons:**
+
+- Users need to be aware of how many workers are in their cluster when setting `steps_per_epoch`. Setting `steps_per_epoch=100` will actually run `100 * W` training batches, which could be confusing to users. Halving the number of workers, without changing any code, would also halve the amount of training performed.
+
+If we go with Option (1), we should disallow batch-level `Callback`s, since in this case `ParameterServerStrategy` with batch-level `Callback`s will always be slower than training on a single machine.
+
+If we go with Option (2) we should support batch-level `Callback`s, but we will use existing logic in `CallbackList` to detect when batch-level `Callback`s are passed, and only incur the performance penalty of syncing workers each batch when the user has passed batch-level `Callback`s (for context, none of Keras's built-in `Callbacks` other than the `ProgressBar` will incur this penalty). This logic was originally put in place to ensure that TPU async mode was only blocked when needed, and applies equally well to `ParameterServerStrategy` without significant modifications. We will also re-use existing logic to log a warning to the user when their batch-level `Callback`s are causing a significant slowdown in training time. This logic also resides in the `CallbackList` object.
```

Comment:
Question @omalleyt12 - does TPU have ProgressBar by default? It looks like it blocks async when verbose=1, which seems to be the default.